protocol-gateway/mqtt-protocol-gateway-template/src/main/java/org/eclipse/hono/gateway/sdk/mqtt2amqp/AbstractMqttProtocolGateway.java

```diff
@@ -0,0 +1,917 @@
+                if (path != null && path.length > 0) {
+                    final Future<Device> authAttempt = authenticateDeviceCertificate(path);
+                    log.debug("authentication with client certificate: {}.",
+                            (authAttempt.succeeded()) ? "succeeded" : "failed");
+                    return authAttempt;
+                }
+            } catch (RuntimeException | SSLPeerUnverifiedException e) {
+                log.debug("could not retrieve client certificate from device endpoint: {}", e.getMessage());
+            }
+        }
+        return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_UNAUTHORIZED));
+    }
+
+    private Future<Device> authenticateWithUsernameAndPassword(final MqttEndpoint endpoint) {
+        final MqttAuth auth = endpoint.auth();
+        if (auth == null || auth.getUsername() == null || auth.getPassword() == null) {
+            return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_UNAUTHORIZED,
+                    "device did not provide credentials in CONNECT packet"));
+        } else {
+            final Future<Device> authenticatedDevice = authenticateDevice(auth.getUsername(), auth.getPassword(),
+                    endpoint.clientIdentifier());
+            if (authenticatedDevice == null) {
+                return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_INTERNAL_ERROR));
+            } else {
+                log.debug("authentication with username/password {}.",
+                        (authenticatedDevice.succeeded()) ? "succeeded" : "failed");
+                return authenticatedDevice;
+            }
+        }
+    }
+
+    private Future<Void> connectGatewayToAmqpAdapter(final Device authenticatedDevice) {
+
+        final String tenantId = authenticatedDevice.getTenantId();
+
+        if (amqpClientConfig.getUsername() != null && amqpClientConfig.getPassword() != null) {
+            return connectGatewayToAmqpAdapter(tenantId, amqpClientConfig);
+        } else {
+            return provideGatewayCredentials(tenantId)
+                    .compose(credentials -> {
+                        final ClientConfigProperties tenantConfig = new ClientConfigProperties(amqpClientConfig);
+                        tenantConfig.setUsername(credentials.getUsername());
+                        tenantConfig.setPassword(credentials.getPassword());
+
+                        return connectGatewayToAmqpAdapter(tenantId, tenantConfig);
+                    });
+        }
+    }
+
+    private Future<Void> connectGatewayToAmqpAdapter(final String tenantId, final ClientConfigProperties clientConfig) {
+
+        final AmqpAdapterClientFactory factory = clientFactoryPerTenant
+                .computeIfAbsent(tenantId, key -> createTenantClientFactory(key, clientConfig));
+
+        return factory.connect() // returns successfully if already connected
+                .map(con -> null);
```

Comment:
When I was running this ProtocolGateway with a wrong AMQP adapter hostname, my first guess was for the corresponding error to come up on startup already. Looking at the code, I see that the connection to the AMQP adapter is only initiated in `handleEndpointConnection()` and that there is one connection per tenant. We may have even discussed this a while back (can't remember exactly), so I understand there are reasons for that.

A downside however is, that if the AMQP adapter is currently unavailable, this `connectGatewayToAmqpAdapter` method return future will not complete since there is by default a continuous retry in establishing the connection.
And any additional `handleEndpointConnection()` invocation for the same tenant, done while the connection attempt is still in progress, will be failed immediately (instead of waiting some time for the ongoing attempt to succeed).

In the protocol adapter code for connecting to the AMQP network, the `connect` invocation is done on startup and when a connection is actually needed as part of a device request, the `checkConnected(waitForCurrentConnectAttemptTimeout)` method is used, containing a timeout. 

Therefore I think this `connectGatewayToAmqpAdapter` method here could be changed to also use a timeout:
````
return factory.isConnected(clientConfig.getConnectTimeout());
````
(a different actual timeout value might be better here, using the connect timeout for simplicity).

The `factory.connect()` invocation could be done in `createTenantClientFactory` (or `createAndConnectTenantClientFactory`), ignoring the return value of `factory.connect()`.
An idea so far, haven't tested it yet. WDYT?

Reply:
`isConnected` check with timeout is used now for all but the first `connectGatewayToAmqpAdapter` invocation for a tenant. Further improvement could be done as part of #5 (see https://github.com/eclipse/hono-extras/issues/5#issuecomment-662253171).